File: backend/src/main/java/org/luxons/sevenwonders/game/resources/Production.java
package org.luxons.sevenwonders.game.resources;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.EnumSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

public class Production {

    private final Resources fixedResources = new Resources();

    private final List<Set<ResourceType>> alternativeResources = new ArrayList<>();

    public void addFixedResource(ResourceType type, int quantity) {
        fixedResources.add(type, quantity);
    }

    public void addChoice(ResourceType... options) {
        EnumSet<ResourceType> optionSet = EnumSet.copyOf(Arrays.asList(options));
        alternativeResources.add(optionSet);
    }

    public void addAll(Resources resources) {
        fixedResources.addAll(resources);
    }

    public void addAll(Production production) {
        fixedResources.addAll(production.getFixedResources());
        alternativeResources.addAll(production.getAlternativeResources());
    }

    public Resources getFixedResources() {
        return fixedResources;
    }

    public List<Set<ResourceType>> getAlternativeResources() {
        return alternativeResources;
    }

    public boolean contains(Resources resources) {
        if (fixedResources.contains(resources)) {
            return true;
        }
        Resources remaining = resources.minus(fixedResources);
        return containedInAlternatives(remaining);
    }

    private boolean containedInAlternatives(Resources resources) {
        return containedInAlternatives(resources, alternativeResources);
    }

    private static boolean containedInAlternatives(Resources resources, List<Set<ResourceType>> alternatives) {
        if (resources.isEmpty()) {
            return true;
        }
        for (ResourceType type : ResourceType.values()) {
            if (resources.getQuantity(type) <= 0) {
                continue;
            }
            Set<ResourceType> candidate = findFirstAlternativeContaining(alternatives, type);
            if (candidate == null) {
                return false; // no alternative produces the resource of this entry
            }
            resources.remove(type, 1);
            alternatives.remove(candidate);
            boolean remainingAreContainedToo = containedInAlternatives(resources, alternatives);
            resources.add(type, 1);
            alternatives.add(candidate);
            if (remainingAreContainedToo) {
                return true;
            }
        }
        return false;
    }

    private static Set<ResourceType> findFirstAlternativeContaining(List<Set<ResourceType>> alternatives,
            ResourceType type) {
        return alternatives.stream().filter(a -> a.contains(type)).findAny().orElse(null);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Production that = (Production) o;
        return Objects.equals(fixedResources, that.fixedResources) && Objects.equals(alternativeResources,
                that.alternativeResources);
    }

    @Override
    public int hashCode() {
        return Objects.hash(fixedResources, alternativeResources);
    }
}

File: sw-engine/src/main/kotlin/org/luxons/sevenwonders/engine/resources/Production.kt
package org.luxons.sevenwonders.engine.resources

import org.luxons.sevenwonders.model.resources.ResourceType
import java.util.EnumSet

data class Production internal constructor(
    private val fixedResources: MutableResources = mutableResourcesOf(),
    // cannot be a Set because the same choices can be there multiple times
    private val alternativeResources: MutableList<Set<ResourceType>> = mutableListOf(),
) {
    fun getFixedResources(): Resources = fixedResources

    fun getAlternativeResources(): List<Set<ResourceType>> = alternativeResources

    fun addFixedResource(type: ResourceType, quantity: Int) = fixedResources.add(type, quantity)

    fun addChoice(vararg options: ResourceType) {
        val optionSet = EnumSet.copyOf(options.toList())
        alternativeResources.add(optionSet)
    }

    fun addAll(resources: Resources) = fixedResources.add(resources)

    fun addAll(production: Production) {
        fixedResources.add(production.fixedResources)
        alternativeResources.addAll(production.getAlternativeResources())
    }

    internal fun asChoices(): List<Set<ResourceType>> {
        val fixedAsChoices = fixedResources.toList().map { EnumSet.of(it) }
        return fixedAsChoices + alternativeResources
    }

    operator fun contains(resources: Resources): Boolean {
        if (fixedResources.containsAll(resources)) {
            return true
        }
        return containedInAlternatives(resources - fixedResources)
    }

    private fun containedInAlternatives(resources: Resources): Boolean =
        containedInAlternatives(resources.toMutableResources(), alternativeResources)

    private fun containedInAlternatives(
        resources: MutableResources,
        alternatives: MutableList<Set<ResourceType>>,
    ): Boolean {
        if (resources.isEmpty()) {
            return true
        }
        for (type in ResourceType.values()) {
            if (resources[type] <= 0) {
                continue
            }
            // return if no alternative produces the resource of this entry
            val candidate = alternatives.firstOrNull { a -> a.contains(type) } ?: return false
            resources.remove(type, 1)
            alternatives.remove(candidate)
            val remainingAreContainedToo = containedInAlternatives(resources, alternatives)
            resources.add(type, 1)
            alternatives.add(candidate)
            if (remainingAreContainedToo) {
                return true
            }
        }
        return false
    }
}
